// react-native-iap/lib/module/types.js
"use strict";

// ============================================================================
// AUTO-GENERATED TYPES — DO NOT EDIT DIRECTLY
// Run `bun run generate:types` after updating any *.graphql schema file.
// ============================================================================

export let ErrorCode = /*#__PURE__*/function (ErrorCode) {
  ErrorCode["ActivityUnavailable"] = "ACTIVITY_UNAVAILABLE";
  ErrorCode["AlreadyOwned"] = "ALREADY_OWNED";
  ErrorCode["AlreadyPrepared"] = "ALREADY_PREPARED";
  ErrorCode["BillingResponseJsonParseError"] = "BILLING_RESPONSE_JSON_PARSE_ERROR";
  ErrorCode["BillingUnavailable"] = "BILLING_UNAVAILABLE";
  ErrorCode["ConnectionClosed"] = "CONNECTION_CLOSED";
  ErrorCode["DeferredPayment"] = "DEFERRED_PAYMENT";
  ErrorCode["DeveloperError"] = "DEVELOPER_ERROR";
  ErrorCode["EmptySkuList"] = "EMPTY_SKU_LIST";
  ErrorCode["FeatureNotSupported"] = "FEATURE_NOT_SUPPORTED";
  ErrorCode["IapNotAvailable"] = "IAP_NOT_AVAILABLE";
  ErrorCode["InitConnection"] = "INIT_CONNECTION";
  ErrorCode["Interrupted"] = "INTERRUPTED";
  ErrorCode["ItemNotOwned"] = "ITEM_NOT_OWNED";
  ErrorCode["ItemUnavailable"] = "ITEM_UNAVAILABLE";
  ErrorCode["NetworkError"] = "NETWORK_ERROR";
  ErrorCode["NotEnded"] = "NOT_ENDED";
  ErrorCode["NotPrepared"] = "NOT_PREPARED";
  ErrorCode["Pending"] = "PENDING";
  ErrorCode["PurchaseError"] = "PURCHASE_ERROR";
  ErrorCode["QueryProduct"] = "QUERY_PRODUCT";
  ErrorCode["ReceiptFailed"] = "RECEIPT_FAILED";
  ErrorCode["ReceiptFinished"] = "RECEIPT_FINISHED";
  ErrorCode["ReceiptFinishedFailed"] = "RECEIPT_FINISHED_FAILED";
  ErrorCode["RemoteError"] = "REMOTE_ERROR";
  ErrorCode["ServiceDisconnected"] = "SERVICE_DISCONNECTED";
  ErrorCode["ServiceError"] = "SERVICE_ERROR";
  ErrorCode["SkuNotFound"] = "SKU_NOT_FOUND";
  ErrorCode["SkuOfferMismatch"] = "SKU_OFFER_MISMATCH";
  ErrorCode["SyncError"] = "SYNC_ERROR";
  ErrorCode["TransactionValidationFailed"] = "TRANSACTION_VALIDATION_FAILED";
  ErrorCode["Unknown"] = "UNKNOWN";
  ErrorCode["UserCancelled"] = "USER_CANCELLED";
  ErrorCode["UserError"] = "USER_ERROR";
  return ErrorCode;
}({});
